Service/AnalyticsFactory.php

```diff
@@ -83,32 +84,106 @@ public function createAnalytics()
             ->setAsyncRequest($this->async && !$this->debug)
             ->setDebug($this->debug);
 
-        if (null !== $request = $this->requestStack->getCurrentRequest()) {
-            $analytics
-                ->setIpOverride($request->getClientIp())
-                ->setUserAgentOverride($request->headers->get('User-Agent', ''));
+        $request = $this->requestStack->getCurrentRequest();
 
-            // Set clientId from "_ga" cookie if exists,
-            // otherwise this must be set at a later point
-            if (null !== $ga = $request->cookies->get('_ga')) {
-                $cookie = $this->parseCookie($ga);
-                $analytics->setClientId(array_pop($cookie));
-            }
+        if (null === $request) {
+            return $analytics;
         }
 
+        $this->handleClientIp($analytics, $request);
+        $this->handleUserAgent($analytics, $request);
+        $this->handleClientId($analytics, $request);
+
         return $analytics;
     }
 
     /**
      * Parse the GA Cookie and return data as an array.
      * Example of GA cookie: _ga:GA1.2.492973748.1449824416.
      *
-     * @param $cookie
+     * @param string $cookie
      *
-     * @return array(version, domainDepth, cid)
+     * @return string[] (version, domainDepth, cid)
      */
     public function parseCookie($cookie)
     {
-        return explode('.', substr($cookie, 2), 3);
+        $parsedCookie = explode('.', substr($cookie, 2), 3);
+        $missingElements = 3 - count($parsedCookie);
+
+        if ($missingElements <= 0) {
+            return $parsedCookie;
+        }
+
+        return array_merge(
+            $parsedCookie,
+            array_fill(0, $missingElements, '')
+        );
+    }
+
+    /**
+     * @param Analytics $analytics
+     * @param Request   $request
+     */
+    private function handleClientIp(Analytics $analytics, Request $request)
+    {
+        $clientIp = $request->getClientIp();
+
+        if (null === $clientIp) {
```

Comment:
Again, is this a real-life scenario? There don't seem to be tests.

Reply:
the signature of method `Request::getClientIp` says:
```
    /**
     * Returns the client IP address.
     *
     * This method can read the client IP address from the "X-Forwarded-For" header
     * when trusted proxies were set via "setTrustedProxies()". The "X-Forwarded-For"
     * header value is a comma+space separated list of IP addresses, the left-most
     * being the original client, and each successive proxy that passed the request
     * adding the IP address where it received the request from.
     *
     * @return string|null The client IP address
     *
     * @see getClientIps()
     * @see http://en.wikipedia.org/wiki/X-Forwarded-For
     */
```

IMHO: it's possible to have a nullable value.